### PR TITLE
Fixes #213

### LINKED ISSUE #213: Non-empty policy when union merging frame-ancestors and default-src
My unit test:
```java
Policy x = Parser.parse("frame-ancestors https://foo.bar", "http://example.com");
Policy y = Parser.parse("default-src 'none'", "http://example.com");
Policy z = Parser.parse("script-src 'self'; img-src 'self'", "http://example.com");
x.union(y);
x.union(z);
assertEquals("frame-ancestors https://foo.bar; default-src; script-src 'self'; img-src 'self'", x.show());
```

**Salvation 2.5.0 - unit test passes:**
```
Expected :frame-ancestors https://foo.bar; default-src; script-src 'self'; img-src 'self'
Actual   :frame-ancestors https://foo.bar; default-src; script-src 'self'; img-src 'self'
```

**Salvation 2.6.0 - unit test fails:**
```
Expected :frame-ancestors https://foo.bar; default-src; script-src 'self'; img-src 'self'
Actual   :frame-ancestors https://foo.bar
```

While in salvation 2.5.0 merge operation seems to be associative, it seems not to be in 2.6.0.

**Salvation 2.5.0:**
`(x ∪ y) ∪ z = x ∪ (y ∪ z)
`

**Salvation 2.6.0:**
`(x ∪ y) ∪ z ≠ x ∪ (y ∪ z)
`

| (x ∪ y) ∪ z |  x ∪ (y ∪ z) |
| ------------- |-------------|
| [cspvalidator.org](https://cspvalidator.org/#headerValue%5B%5D=frame-ancestors+https%3A%2F%2Ffoo.bar&headerValue%5B%5D=default-src+'none'&headerValue%5B%5D=script-src+'self'%3B+img-src+'self'&strategy=union) | [cspvalidator.org](https://cspvalidator.org/#headerValue%5B%5D=default-src+'none'&headerValue%5B%5D=frame-ancestors+https%3A%2F%2Ffoo.bar&headerValue%5B%5D=script-src+'self'%3B+img-src+'self'&strategy=union) |

_Originally posted by @cnsgithub in https://github.com/shapesecurity/salvation/issues/210#issuecomment-436885547_

### PR DESCRIPTION
**Pull request by shekyan**
and tests commutativity

### CODE REVIEW

File: src/main/java/com/shapesecurity/salvation/data/Policy.java [hunk 1]
```diff
@@ -52,41 +51,53 @@ public void setOrigin(@Nonnull Origin origin) {
...
+		for (Map.Entry<Class<?>, Directive<? extends DirectiveValue>> entry : other.directives.entrySet()) {
+			this.unionDirectivePrivate(entry.getValue());
+		}
+		this.directives.entrySet().removeIf(entry -> !other.directives.containsKey(entry.getKey()));
 
 		this.optimise();
+		other.optimise();
+	}
+
+	private static void checkForMergeValidity(@Nonnull Policy p) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/data/Policy.java]: We should probably turn this into an instance method now.

File: src/main/java/com/shapesecurity/salvation/data/Policy.java [hunk 2]
```diff
@@ -52,41 +51,53 @@ public void setOrigin(@Nonnull Origin origin) {
...
+		for (Map.Entry<Class<?>, Directive<? extends DirectiveValue>> entry : other.directives.entrySet()) {
+			this.intersectDirectivePrivate(entry.getValue());
 		}
 
-		if (this.directives.containsKey(ReferrerDirective.class) || other.directives
-				.containsKey(ReferrerDirective.class)) {
-			throw new IllegalArgumentException(
-					"Cannot merge policies if either policy contains a referrer directive.");
-		}
+		this.optimise();
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/data/Policy.java]: Add `other.optimise()` here.

File: src/main/java/com/shapesecurity/salvation/data/Policy.java (hunk 1 above)
Comment (shekyan) [reply src/main/java/com/shapesecurity/salvation/data/Policy.java]: done.

File: src/main/java/com/shapesecurity/salvation/data/Policy.java (hunk 2 above)
Comment (shekyan) [reply src/main/java/com/shapesecurity/salvation/data/Policy.java]: done.

File: src/test/java/com/shapesecurity/salvation/PolicyMergeTest.java [hunk 3]
```diff
@@ -645,7 +656,7 @@ public void testUnionNone() {
 		p = parse("frame-ancestors 'none' 'none'");
 		q = parse("frame-ancestors 'self'");
 		p.union(q);
-		assertEquals("frame-ancestors 'self'", p.show());
+		assertEquals("", p.show());
```
[the remark this answers is not in the document]
Comment (shekyan) [reply src/test/java/com/shapesecurity/salvation/PolicyMergeTest.java]: I would think so too, but `frame-ancestors 'none' 'none'` contains invalid source-list thus invalidates the directive. Union merging empty policy with anything produces empty policy.

[the remark this answers is not in the document]
Comment (shekyan) [reply src/test/java/com/shapesecurity/salvation/PolicyMergeTest.java]: on the other hand, per https://w3c.github.io/webappsec-csp/#parse-serialized-policy step 7, invalid source-list produces an empty `directive-value`, which brings us to recent discussion. WDYT?

[the remark this answers is not in the document]
Comment (michaelficarra) [reply src/test/java/com/shapesecurity/salvation/PolicyMergeTest.java]: Oh wow, you're right. This relies upon https://github.com/w3c/webappsec-csp/pull/363, which intentionally *doesn't* use `frame-ancestors 'none'` in a malformed policy. I'm fine with getting this in as-is, but we need to follow up with an implementation of https://github.com/w3c/webappsec-csp/pull/363 and possibly further discussion about `frame-ancestors`.

**Pull request merged**